src/compiler/csharp_generator.cc

```diff
@@ -323,17 +323,50 @@ std::vector<const Descriptor*> GetUsedMessages(
   return result;
 }
 
-void GenerateMarshallerFields(Printer* out, const ServiceDescriptor* service) {
+void GenerateMarshallerFields(Printer* out, const ServiceDescriptor* service,
+                              bool use_buffer_serialization) {
   std::vector<const Descriptor*> used_messages = GetUsedMessages(service);
-  for (size_t i = 0; i < used_messages.size(); i++) {
-    const Descriptor* message = used_messages[i];
+  if (use_buffer_serialization) {
+    // Generate buffer serialization marshallers
```

Comment:
Not sure if you removed the #if by accident or not, but we still want to have `#if GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION` to allow people to use IBufferMessage and target old .NET standard versions.

Reply:
I guess to support the `#if` we would need to output both code generation when it is enabled.

OR!, we could have the `#if` inside the helper method. And have a helper method for reading content as well

Reply:
Would this work?

```csharp
private static void __Helper_WriteMessage(T message, grpc::SerializationContext context) where T : global::Google.Protobuf.IMessage
{
    #if GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION
    global::Google.Protobuf.IBufferMessage bufferMessage = message as global::Google.Protobuf.IBufferMessage;
    if (bufferMessage != null)
    {
        global::Google.Protobuf.CodedOutputWriter writer = new global::Google.Protobuf.CodedOutputWriter(context.GetBufferWriter());
        arg.WriteTo(ref writer);
        writer.Flush();
        context.Complete();
        return;
    }
    #endif
    context.Complete(global::Google.Protobuf.MessageExtensions.ToByteArray(message));
}
```

It would mean codegen only writes the marshaller definitions once.

We could then have `__Helper_ReadMessage` with its own `#if` check.

Reply:
```csharp
private static T __Helper_ReadMessage<T>(grpc::DeserializationContext context, global::Google.Protobuf.MessageParser<T> parser) where T : global::Google.Protobuf.IMessage<T>
{
    #if GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION
    return parser.ParseFrom(context.PayloadAsReadOnlySequence());
    #else
    return parser.ParseFrom(context.PayloadAsNewBuffer());
    #endif
}
```

Reply:
Other than I think you reversed the condition in #if, yes it works.